pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -681,12 +681,35 @@ pub mod pallet {
 					sum.saturating_add(T::Fungibles::total_issuance(id))
 				});
 
-			let amount = burnt
-				.checked_mul(&total_collateral_issuance)
-				.ok_or(ArithmeticError::Overflow)? // TODO: do we need a fallback if this fails?
+			defensive_assert!(
+				sum_of_issuances >= burnt,
+				"burnt amount exceeds the total supply of all bonded currencies"
+			);
+
+			// divide first to avoid overflow, remainder will be handled seperately
+			let divisible_part = total_collateral_issuance
+				.checked_div(&sum_of_issuances)
+				.ok_or(Error::<T>::Internal)? // because sum_of_issuances >= burnt > 0, this is theoretically impossible
+				.checked_mul(&burnt)
+				// An overflow is also theoretically impossible, as burnt <= sum_of_issuances
+				.ok_or(Error::<T>::Internal)?;
+
+			// for the remainder, we multiply first and divide next
+			let remainder_part = total_collateral_issuance
+				.checked_rem(&sum_of_issuances)
+				.ok_or(Error::<T>::Internal)? // Should only fail on sum_of_issuances == 0, same as above
+				.checked_mul(&burnt)
+				// This can indeed overflow if (sum_of_issuances - 1) * burnt > capacity(u128)
+				.ok_or(ArithmeticError::Overflow)?
```

Comment:
Very large sum_of_issuances make large remainders more likely, which can in turn result in a product that too large.
We can accept that refunds would sometimes fail, which would require accounts with lower balances to refund first (after which the situation may have changed).
Or we can look for fallback strategies, such as:
- saturating multiplication, which could mean that people receive only a fraction of the funds they would be entitled to
- scaling, which sacrifices precision
- partial refunds, where we only burn the amount of funds for which we can properly compute the collateral to be refunded, and leave the rest untouched.

Reply:
With https://github.com/KILTprotocol/kilt-node/commit/08b71c4c8966febb0c17df070b7b3e941f77115b overflow issues are now a thing of the past.